// openclaw/dist/plugin-sdk/browser-node-runtime.js
import { n as defaultRuntime } from "../runtime-BXaxArxm.js";
import { i as runExec } from "../exec-uTlaAO3l.js";
import { a as isLoopbackHost } from "../net-Gtz838RS.js";
import { o as resolveGatewayAuth } from "../auth-CMmjOHxZ.js";
import { dn as ErrorCodes, fn as errorShape } from "../method-scopes-Cgh_6px_.js";
import { t as rawDataToString } from "../ws-mJNtsIhg.js";
import { t as withTimeout } from "../with-timeout-C1DfH8XT.js";
import { t as startLazyPluginServiceModule } from "../lazy-service-module-CJYq7qVi.js";
import { n as callGatewayFromCli, t as addGatewayClientOptions } from "../gateway-rpc-CZqo2-ji.js";
import { n as runCommandWithRuntime } from "../cli-utils-CUiAJaWR.js";
import { i as resolveNodeCommandAllowlist, n as isNodeCommandAllowed } from "../node-command-policy-CZnSinzk.js";
import { a as respondUnavailableOnNodeInvokeError, s as safeParseJson, t as ensureGatewayStartupAuth } from "../startup-auth-CzulTTBe.js";
import "../browser-node-runtime-uhg_sLaE.js";
export { ErrorCodes, addGatewayClientOptions, callGatewayFromCli, defaultRuntime, ensureGatewayStartupAuth, errorShape, isLoopbackHost, isNodeCommandAllowed, rawDataToString, resolveGatewayAuth, resolveNodeCommandAllowlist, respondUnavailableOnNodeInvokeError, runCommandWithRuntime, runExec, safeParseJson, startLazyPluginServiceModule, withTimeout };
