// openclaw/dist/plugin-sdk/tlon.js
import { t as formatDocsLink } from "../links-BFFMr2UK.js";
import { n as normalizeAccountId, t as DEFAULT_ACCOUNT_ID } from "../account-id-BfovDWrj.js";
import { r as buildChannelConfigSchema } from "../config-schema-Bdf-zEgr.js";
import { l as patchScopedAccountConfig, t as applyAccountNameToChannelSection } from "../setup-helpers-CwEENaWo.js";
import { t as createDedupeCache } from "../dedupe-MmEr1T7f.js";
import { n as fetchWithSsrFGuard } from "../fetch-guard-DDS3GI-I.js";
import { s as isBlockedHostnameOrIp, t as SsrFBlockedError } from "../ssrf-C2YYZEse.js";
import { n as emptyPluginConfigSchema } from "../config-schema-C7tGFgOZ.js";
import { t as createChannelReplyPipeline } from "../channel-reply-pipeline-tHqXP07h.js";
import { r as buildComputedAccountStatusSnapshot } from "../status-helpers-wqet6BlE.js";
import { t as createLoggerBackedRuntime } from "../runtime-logger-DIAymU6t.js";
import "../runtime-B3RuX90K.js";
import { t as createOptionalChannelSetupSurface } from "../channel-setup-BRxI_ucP.js";
//#region src/plugin-sdk/tlon.ts
const tlonSetup = createOptionalChannelSetupSurface({
	channel: "tlon",
	label: "Tlon",
	npmSpec: "@openclaw/tlon",
	docsPath: "/channels/tlon"
});
const tlonSetupAdapter = tlonSetup.setupAdapter;
const tlonSetupWizard = tlonSetup.setupWizard;
//#endregion
export { DEFAULT_ACCOUNT_ID, SsrFBlockedError, applyAccountNameToChannelSection, buildChannelConfigSchema, buildComputedAccountStatusSnapshot, createChannelReplyPipeline, createDedupeCache, createLoggerBackedRuntime, emptyPluginConfigSchema, fetchWithSsrFGuard, formatDocsLink, isBlockedHostnameOrIp, normalizeAccountId, patchScopedAccountConfig, tlonSetupAdapter, tlonSetupWizard };
