// openclaw/dist/api-HwawTrNY.js
import { d as readStringValue } from "./string-coerce-CIT1-v_a.js";
import { r as normalizeProviderId } from "./provider-id-C2ErhgDG.js";
import { t as resolveProviderEndpoint } from "./provider-attribution-CsssnIri.js";
import "./provider-tools-CdoMvJZX.js";
import { n as getModelProviderHint } from "./provider-model-shared-CJJooY3e.js";
import "./text-runtime-dYMrRdFH.js";
import "./model-definitions-CYIm1_75.js";
import "./provider-catalog-DPhJGhUN.js";
import "./onboard-B17kmVwy.js";
import "./provider-models-BOaxCIjZ.js";
//#region extensions/xai/api.ts
function isXaiNativeEndpoint(baseUrl) {
	return typeof baseUrl === "string" && resolveProviderEndpoint(baseUrl).endpointClass === "xai-native";
}
function isXaiModelHint(modelId) {
	return getModelProviderHint(modelId) === "x-ai";
}
function shouldUseXaiResponsesTransport(params) {
	if (params.api !== "openai-completions") return false;
	if (isXaiNativeEndpoint(params.baseUrl)) return true;
	return normalizeProviderId(params.provider) === "xai" && !params.baseUrl;
}
function shouldContributeXaiCompat(params) {
	if (params.model.api !== "openai-completions") return false;
	return isXaiNativeEndpoint(params.model.baseUrl) || isXaiModelHint(params.modelId);
}
function resolveXaiTransport(params) {
	if (!shouldUseXaiResponsesTransport(params)) return;
	return {
		api: "openai-responses",
		baseUrl: readStringValue(params.baseUrl)
	};
}
//#endregion
export { resolveXaiTransport as n, shouldContributeXaiCompat as r, isXaiModelHint as t };
