// openclaw/dist/extensionAPI.js
import { S as resolveThinkingDefault } from "./model-selection-JHakuCn9.js";
import { a as resolveAgentDir, m as resolveAgentWorkspaceDir } from "./agent-scope-BfV3ML7Z.js";
import { d as ensureAgentWorkspace } from "./workspace-BPG_m9MI.js";
import { n as DEFAULT_MODEL, r as DEFAULT_PROVIDER } from "./defaults-CF547YAW.js";
import { o as saveSessionStore } from "./store-ZWCZ_LZI.js";
import "./sessions-DmA78qkx.js";
import { i as resolveSessionFilePath, u as resolveStorePath } from "./paths-DgoT9gfU.js";
import { t as loadSessionStore } from "./store-load-DlllQPOX.js";
import { n as resolveAgentIdentity } from "./identity-D_zFlCIB.js";
import { t as resolveAgentTimeoutMs } from "./timeout-CCmjZeel.js";
import { t as runEmbeddedPiAgent } from "./pi-embedded-runner-CRRqVG8f.js";
import "./pi-embedded-BQ8E8Nhk.js";
//#region src/extensionAPI.ts
if (process.env.VITEST !== "true" && process.env.OPENCLAW_SUPPRESS_EXTENSION_API_WARNING !== "1") process.emitWarning("openclaw/extension-api is deprecated. Migrate to api.runtime.agent.* or focused openclaw/plugin-sdk/<subpath> imports. See https://docs.openclaw.ai/plugins/sdk-migration", {
	code: "OPENCLAW_EXTENSION_API_DEPRECATED",
	detail: "This compatibility bridge is temporary. Bundled plugins should use the injected plugin runtime instead of importing host-side agent helpers directly. Migration guide: https://docs.openclaw.ai/plugins/sdk-migration"
});
//#endregion
export { DEFAULT_MODEL, DEFAULT_PROVIDER, ensureAgentWorkspace, loadSessionStore, resolveAgentDir, resolveAgentIdentity, resolveAgentTimeoutMs, resolveAgentWorkspaceDir, resolveSessionFilePath, resolveStorePath, resolveThinkingDefault, runEmbeddedPiAgent, saveSessionStore };
