// openclaw/dist/googlechat-B8eE-oDM.js
import "./message-channel-DKMzUOS9.js";
import "./types.secrets-D-R9X_jm.js";
import "./config-schema-Bdf-zEgr.js";
import "./setup-helpers-CwEENaWo.js";
import "./zod-schema.providers-core-BaNrkHru.js";
import "./fetch-guard-DDS3GI-I.js";
import "./common-ub5s4Xdi.js";
import { n as resolveChannelGroupRequireMention } from "./group-policy-BOK9r2Js.js";
import "./fetch-CraQ-PKa.js";
import "./channel-policy-B3TuWXAd.js";
import "./dm-policy-shared-1zVsdEGf.js";
import "./setup-wizard-helpers-np_EArk_.js";
import "./channel-reply-pipeline-tHqXP07h.js";
import "./channel-pairing-B4BhGbz7.js";
import "./status-helpers-wqet6BlE.js";
import "./webhook-ingress-lquZa6ui.js";
import { t as createOptionalChannelSetupSurface } from "./channel-setup-BRxI_ucP.js";
import "./web-media-Bh3YWiOa.js";
import "./outbound-media-CVfLFyGl.js";
//#region src/plugin-sdk/googlechat.ts
function resolveGoogleChatGroupRequireMention(params) {
	return resolveChannelGroupRequireMention({
		cfg: params.cfg,
		channel: "googlechat",
		groupId: params.groupId,
		accountId: params.accountId
	});
}
const googlechatSetup = createOptionalChannelSetupSurface({
	channel: "googlechat",
	label: "Google Chat",
	npmSpec: "@openclaw/googlechat",
	docsPath: "/channels/googlechat"
});
const googlechatSetupAdapter = googlechatSetup.setupAdapter;
const googlechatSetupWizard = googlechatSetup.setupWizard;
//#endregion
export { googlechatSetupWizard as n, resolveGoogleChatGroupRequireMention as r, googlechatSetupAdapter as t };
