// openclaw/dist/infra-runtime-DzGODV0M.js
import "./errors-DcAstocv.js";
import "./tmp-openclaw-dir-CWQcmOLf.js";
import "./env-WgjbeXMN.js";
import "./file-lock-CZ6aL3dF.js";
import "./undici-global-dispatcher-BQ64ogHN.js";
import "./fetch-guard-DDS3GI-I.js";
import "./ssrf-C2YYZEse.js";
import "./exec-approval-reply-B4ng3Kgc.js";
import "./exec-approvals-DT4-ZzBV.js";
import "./fs-safe-DQfR1Er7.js";
import "./system-events-Bc8HfhQ3.js";
import { n as drainPendingDeliveries } from "./delivery-queue-wYNXw79t.js";
import "./proxy-fetch-BIsw6Ncw.js";
import "./retry-DcseOnly.js";
import "./secret-file-BAKR3SJz.js";
import "./http-body-Dittl_fd.js";
import "./approval-native-runtime-C_-fdhx0.js";
import "./exec-approval-session-target-BKerYSxw.js";
import "./heartbeat-visibility-vJxEaSRp.js";
import "./transport-ready-1f0Fjmoo.js";
import "./identity-BdZU5rK9.js";
import "./retry-policy-DIzHfSBh.js";
import "./ssrf-policy-DANBiMfe.js";
//#region src/plugin-sdk/infra-runtime.ts
function normalizeWhatsAppReconnectAccountId(accountId) {
	return (accountId ?? "").trim() || "default";
}
const WHATSAPP_NO_LISTENER_ERROR_RE = /No active WhatsApp Web listener/i;
/**
* @deprecated Prefer plugin-owned reconnect policy wired through
* `drainPendingDeliveries(...)`. This compatibility shim preserves the
* historical public SDK symbol for existing plugin callers.
*/
async function drainReconnectQueue(opts) {
	const normalizedAccountId = normalizeWhatsAppReconnectAccountId(opts.accountId);
	await drainPendingDeliveries({
		drainKey: `whatsapp:${normalizedAccountId}`,
		logLabel: "WhatsApp reconnect drain",
		cfg: opts.cfg,
		log: opts.log,
		stateDir: opts.stateDir,
		deliver: opts.deliver,
		selectEntry: (entry) => ({
			match: entry.channel === "whatsapp" && normalizeWhatsAppReconnectAccountId(entry.accountId) === normalizedAccountId && typeof entry.lastError === "string" && WHATSAPP_NO_LISTENER_ERROR_RE.test(entry.lastError),
			bypassBackoff: true
		})
	});
}
//#endregion
export { drainReconnectQueue as t };
