// openclaw/dist/library-D7Ih3S7l.js
import { u as normalizeE164 } from "./utils-Supr9KvU.js";
import { a as loadConfig } from "./io-BxA9F2P4.js";
import "./config-BgwTLDRu.js";
import { o as saveSessionStore } from "./store-ZWCZ_LZI.js";
import { u as resolveStorePath } from "./paths-DgoT9gfU.js";
import { n as resolveSessionKey, t as deriveSessionKey } from "./session-key-CA8qiuFn.js";
import { t as loadSessionStore } from "./store-load-DlllQPOX.js";
import { i as handlePortError, n as describePortOwner, r as ensurePortAvailable, t as PortInUseError } from "./ports-CF9KS2uR.js";
import { t as applyTemplate } from "./templating-QtWUDFBt.js";
import { t as createDefaultDeps } from "./deps-Dh9KhZEt.js";
import { t as waitForever } from "./wait-BLA-e14P.js";
//#region src/library.ts
let replyRuntimePromise = null;
let promptRuntimePromise = null;
let binariesRuntimePromise = null;
let execRuntimePromise = null;
let webChannelRuntimePromise = null;
function loadReplyRuntime() {
	replyRuntimePromise ??= import("./reply.runtime-D9upoYr2.js");
	return replyRuntimePromise;
}
function loadPromptRuntime() {
	promptRuntimePromise ??= import("./prompt-DGMFEtct.js");
	return promptRuntimePromise;
}
function loadBinariesRuntime() {
	binariesRuntimePromise ??= import("./binaries-Dq50M9gt.js");
	return binariesRuntimePromise;
}
function loadExecRuntime() {
	execRuntimePromise ??= import("./exec-LmooGFld.js");
	return execRuntimePromise;
}
function loadWebChannelRuntime() {
	webChannelRuntimePromise ??= import("./runtime-web-channel-plugin-Bk3scS8B.js");
	return webChannelRuntimePromise;
}
const getReplyFromConfig = async (...args) => (await loadReplyRuntime()).getReplyFromConfig(...args);
const promptYesNo = async (...args) => (await loadPromptRuntime()).promptYesNo(...args);
const ensureBinary = async (...args) => (await loadBinariesRuntime()).ensureBinary(...args);
const runExec = async (...args) => (await loadExecRuntime()).runExec(...args);
const runCommandWithTimeout = async (...args) => (await loadExecRuntime()).runCommandWithTimeout(...args);
const monitorWebChannel = async (...args) => (await loadWebChannelRuntime()).monitorWebChannel(...args);
//#endregion
export { PortInUseError, applyTemplate, createDefaultDeps, deriveSessionKey, describePortOwner, ensureBinary, ensurePortAvailable, getReplyFromConfig, handlePortError, loadConfig, loadSessionStore, monitorWebChannel, normalizeE164, promptYesNo, resolveSessionKey, resolveStorePath, runCommandWithTimeout, runExec, saveSessionStore, waitForever };
