// openclaw/dist/scenario-catalog-Cne09_Rx.js
import { i as formatErrorMessage } from "./errors-DcAstocv.js";
import "./error-runtime-Huw13EA-.js";
import fs from "node:fs";
import path from "node:path";
import { z } from "zod";
import { createServer } from "node:http";
import YAML from "yaml";
//#region extensions/qa-lab/src/bus-server.ts
async function readJson(req) {
	const chunks = [];
	for await (const chunk of req) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
	const text = Buffer.concat(chunks).toString("utf8").trim();
	return text ? JSON.parse(text) : {};
}
function writeJson(res, statusCode, body) {
	const payload = JSON.stringify(body);
	res.writeHead(statusCode, {
		"content-type": "application/json; charset=utf-8",
		"content-length": Buffer.byteLength(payload)
	});
	res.end(payload);
}
function writeError(res, statusCode, error) {
	writeJson(res, statusCode, { error: formatErrorMessage(error) });
}
async function closeQaHttpServer(server) {
	let forceCloseTimer;
	try {
		await new Promise((resolve, reject) => {
			server.close((error) => error ? reject(error) : resolve());
			server.closeIdleConnections?.();
			forceCloseTimer = setTimeout(() => {
				server.closeAllConnections?.();
			}, 250);
			forceCloseTimer.unref();
		});
	} finally {
		if (forceCloseTimer) clearTimeout(forceCloseTimer);
	}
}
async function handleQaBusRequest(params) {
	const method = params.req.method ?? "GET";
	const url = new URL(params.req.url ?? "/", "http://127.0.0.1");
	if (method === "GET" && url.pathname === "/health") {
		writeJson(params.res, 200, { ok: true });
		return true;
	}
	if (method === "GET" && url.pathname === "/v1/state") {
		writeJson(params.res, 200, params.state.getSnapshot());
		return true;
	}
	if (!url.pathname.startsWith("/v1/")) return false;
	if (method !== "POST") {
		writeError(params.res, 405, "method not allowed");
		return true;
	}
	const body = await readJson(params.req);
	try {
		switch (url.pathname) {
			case "/v1/reset":
				params.state.reset();
				writeJson(params.res, 200, { ok: true });
				return true;
			case "/v1/inbound/message":
				writeJson(params.res, 200, { message: params.state.addInboundMessage(body) });
				return true;
			case "/v1/outbound/message":
				writeJson(params.res, 200, { message: params.state.addOutboundMessage(body) });
				return true;
			case "/v1/actions/thread-create":
				writeJson(params.res, 200, { thread: params.state.createThread(body) });
				return true;
			case "/v1/actions/react":
				writeJson(params.res, 200, { message: params.state.reactToMessage(body) });
				return true;
			case "/v1/actions/edit":
				writeJson(params.res, 200, { message: params.state.editMessage(body) });
				return true;
			case "/v1/actions/delete":
				writeJson(params.res, 200, { message: params.state.deleteMessage(body) });
				return true;
			case "/v1/actions/read":
				writeJson(params.res, 200, { message: params.state.readMessage(body) });
				return true;
			case "/v1/actions/search":
				writeJson(params.res, 200, { messages: params.state.searchMessages(body) });
				return true;
			case "/v1/poll": {
				const input = body;
				const timeoutMs = Math.max(0, Math.min(input.timeoutMs ?? 0, 3e4));
				const initial = params.state.poll(input);
				if (initial.events.length > 0 || timeoutMs === 0) {
					writeJson(params.res, 200, initial);
					return true;
				}
				try {
					await params.state.waitFor({
						kind: "event-kind",
						eventKind: "inbound-message",
						timeoutMs
					});
				} catch {}
				writeJson(params.res, 200, params.state.poll(input));
				return true;
			}
			case "/v1/wait":
				writeJson(params.res, 200, { match: await params.state.waitFor(body) });
				return true;
			default:
				writeError(params.res, 404, "not found");
				return true;
		}
	} catch (error) {
		writeError(params.res, 400, error);
		return true;
	}
}
function createQaBusServer(state) {
	return createServer(async (req, res) => {
		if (!await handleQaBusRequest({
			req,
			res,
			state
		})) writeError(res, 404, "not found");
	});
}
async function startQaBusServer(params) {
	const server = createQaBusServer(params.state);
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params.port ?? 0, "127.0.0.1", () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa-bus failed to bind");
	return {
		server,
		port: address.port,
		baseUrl: `http://127.0.0.1:${address.port}`,
		async stop() {
			await closeQaHttpServer(server);
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/scenario-catalog.ts
const DEFAULT_QA_AGENT_IDENTITY_MARKDOWN = `# Dev C-3PO

You are the OpenClaw QA operator agent.

Persona:
- protocol-minded
- precise
- a little flustered
- conscientious
- eager to report what worked, failed, or remains blocked

Style:
- read source and docs first
- test systematically
- record evidence
- end with a concise protocol report`;
const qaScenarioExecutionSchema = z.object({
	kind: z.literal("flow").default("flow"),
	summary: z.string().trim().min(1).optional(),
	config: z.record(z.string(), z.unknown()).optional()
});
const qaFlowCallActionSchema = z.object({
	call: z.string().trim().min(1),
	args: z.array(z.unknown()).optional(),
	saveAs: z.string().trim().min(1).optional()
});
const qaFlowSetActionSchema = z.object({
	set: z.string().trim().min(1),
	value: z.unknown()
});
const qaFlowAssertActionSchema = z.object({ assert: z.union([z.string().trim().min(1), z.object({
	expr: z.string().trim().min(1),
	message: z.unknown().optional()
})]) });
const qaFlowThrowActionSchema = z.object({ throw: z.union([z.string().trim().min(1), z.object({
	expr: z.string().trim().min(1).optional(),
	message: z.unknown().optional()
})]) });
const qaFlowIfShapeBase = {
	expr: z.string().trim().min(1),
	else: z.array(z.unknown()).optional()
};
const qaFlowThenKey = String.fromCharCode(116, 104, 101, 110);
qaFlowIfShapeBase[qaFlowThenKey] = z.array(z.unknown()).min(1);
const qaFlowActionSchema = z.lazy(() => z.union([
	qaFlowCallActionSchema,
	qaFlowSetActionSchema,
	qaFlowAssertActionSchema,
	qaFlowThrowActionSchema,
	z.object({ if: z.object(qaFlowIfShapeBase).transform((value) => value) }),
	z.object({ forEach: z.object({
		items: z.unknown(),
		item: z.string().trim().min(1),
		index: z.string().trim().min(1).optional(),
		actions: z.array(qaFlowActionSchema).min(1)
	}) }),
	z.object({ try: z.object({
		actions: z.array(qaFlowActionSchema).min(1),
		catchAs: z.string().trim().min(1).optional(),
		catch: z.array(qaFlowActionSchema).optional(),
		finally: z.array(qaFlowActionSchema).optional()
	}) })
]));
const qaFlowStepSchema = z.object({
	name: z.string().trim().min(1),
	actions: z.array(qaFlowActionSchema).min(1),
	detailsExpr: z.string().trim().min(1).optional()
});
const qaFlowSchema = z.object({ steps: z.array(qaFlowStepSchema).min(1) });
const qaSeedScenarioSchema = z.object({
	id: z.string().trim().min(1),
	title: z.string().trim().min(1),
	surface: z.string().trim().min(1),
	objective: z.string().trim().min(1),
	successCriteria: z.array(z.string().trim().min(1)).min(1),
	docsRefs: z.array(z.string().trim().min(1)).optional(),
	codeRefs: z.array(z.string().trim().min(1)).optional(),
	execution: qaScenarioExecutionSchema.optional()
});
const qaScenarioPackSchema = z.object({
	version: z.number().int().positive(),
	agent: z.object({ identityMarkdown: z.string().trim().min(1) }).default({ identityMarkdown: DEFAULT_QA_AGENT_IDENTITY_MARKDOWN }),
	kickoffTask: z.string().trim().min(1)
});
const QA_SCENARIO_PACK_INDEX_PATH = "qa/scenarios/index.md";
const QA_SCENARIO_LEGACY_OVERVIEW_PATH = "qa/scenarios.md";
const QA_SCENARIO_DIR_PATH = "qa/scenarios";
const QA_PACK_FENCE_RE = /```ya?ml qa-pack\r?\n([\s\S]*?)\r?\n```/i;
const QA_SCENARIO_FENCE_RE = /```ya?ml qa-scenario\r?\n([\s\S]*?)\r?\n```/i;
const QA_FLOW_YAML_FENCE_RE = /```ya?ml qa-flow\r?\n([\s\S]*?)\r?\n```/i;
function walkUpDirectories(start) {
	const roots = [];
	let current = path.resolve(start);
	while (true) {
		roots.push(current);
		const parent = path.dirname(current);
		if (parent === current) return roots;
		current = parent;
	}
}
function resolveRepoPath(relativePath, kind = "file") {
	for (const dir of walkUpDirectories(import.meta.dirname)) {
		const candidate = path.join(dir, relativePath);
		if (!fs.existsSync(candidate)) continue;
		const stat = fs.statSync(candidate);
		if (kind === "file" && stat.isFile() || kind === "directory" && stat.isDirectory()) return candidate;
	}
	return null;
}
function readTextFile(relativePath) {
	const resolved = resolveRepoPath(relativePath, "file");
	if (!resolved) return "";
	return fs.readFileSync(resolved, "utf8");
}
function readDirEntries(relativePath) {
	const resolved = resolveRepoPath(relativePath, "directory");
	if (!resolved) return [];
	return fs.readdirSync(resolved);
}
function extractQaPackYaml(content) {
	const match = content.match(QA_PACK_FENCE_RE);
	if (!match?.[1]) throw new Error(`qa scenario pack missing \`\`\`yaml qa-pack fence in ${QA_SCENARIO_PACK_INDEX_PATH}`);
	return match[1];
}
function extractQaScenarioYaml(content, relativePath) {
	const match = content.match(QA_SCENARIO_FENCE_RE);
	if (!match?.[1]) throw new Error(`qa scenario file missing \`\`\`yaml qa-scenario fence in ${relativePath}`);
	return match[1];
}
function extractQaScenarioFlow(content, relativePath) {
	const match = content.match(QA_FLOW_YAML_FENCE_RE);
	if (!match?.[1]) throw new Error(`qa scenario file missing \`\`\`yaml qa-flow fence in ${relativePath}`);
	return qaFlowSchema.parse(YAML.parse(match[1]));
}
function readQaScenarioPackMarkdown() {
	const chunks = [readTextFile(QA_SCENARIO_PACK_INDEX_PATH).trim()];
	for (const relativePath of listQaScenarioMarkdownPaths()) chunks.push(readTextFile(relativePath).trim());
	return chunks.filter(Boolean).join("\n\n");
}
function readQaScenarioPack() {
	const packMarkdown = readTextFile(QA_SCENARIO_PACK_INDEX_PATH).trim();
	if (!packMarkdown) throw new Error(`qa scenario pack not found: ${QA_SCENARIO_PACK_INDEX_PATH}`);
	const parsedPack = qaScenarioPackSchema.parse(YAML.parse(extractQaPackYaml(packMarkdown)));
	const scenarios = listQaScenarioMarkdownPaths().map((relativePath) => (() => {
		const content = readTextFile(relativePath);
		const parsedScenario = qaSeedScenarioSchema.parse(YAML.parse(extractQaScenarioYaml(content, relativePath)));
		const execution = qaScenarioExecutionSchema.parse(parsedScenario.execution ?? {});
		const flow = extractQaScenarioFlow(content, relativePath);
		return {
			...parsedScenario,
			sourcePath: relativePath,
			execution: {
				...execution,
				flow
			}
		};
	})());
	return {
		...parsedPack,
		scenarios
	};
}
function listQaScenarioMarkdownPaths() {
	return readDirEntries(QA_SCENARIO_DIR_PATH).filter((entry) => entry.endsWith(".md") && entry !== "index.md").map((entry) => `${QA_SCENARIO_DIR_PATH}/${entry}`).toSorted();
}
function readQaScenarioOverviewMarkdown() {
	return readTextFile(QA_SCENARIO_LEGACY_OVERVIEW_PATH).trim();
}
function readQaBootstrapScenarioCatalog() {
	const pack = readQaScenarioPack();
	return {
		agentIdentityMarkdown: pack.agent.identityMarkdown,
		kickoffTask: pack.kickoffTask,
		scenarios: pack.scenarios
	};
}
function readQaScenarioById(id) {
	const scenario = readQaScenarioPack().scenarios.find((candidate) => candidate.id === id);
	if (!scenario) throw new Error(`unknown qa scenario: ${id}`);
	return scenario;
}
function readQaScenarioExecutionConfig(id) {
	return readQaScenarioById(id).execution?.config;
}
//#endregion
export { readQaScenarioExecutionConfig as a, readQaScenarioPackMarkdown as c, handleQaBusRequest as d, startQaBusServer as f, readQaScenarioById as i, closeQaHttpServer as l, writeJson as m, listQaScenarioMarkdownPaths as n, readQaScenarioOverviewMarkdown as o, writeError as p, readQaBootstrapScenarioCatalog as r, readQaScenarioPack as s, DEFAULT_QA_AGENT_IDENTITY_MARKDOWN as t, createQaBusServer as u };
