// openclaw/dist/sessions-DmA78qkx.js
import { a as loadConfig } from "./io-BxA9F2P4.js";
import "./store-ZWCZ_LZI.js";
import { i as resolveMainSessionKey } from "./main-session-DIxrImRz.js";
import { u as resolveStorePath } from "./paths-DgoT9gfU.js";
import "./reset-B1Ek-7XP.js";
import "./session-key-CA8qiuFn.js";
import { t as deliveryContextFromSession } from "./delivery-context-ymOZ0DxO.js";
import { t as loadSessionStore } from "./store-load-DlllQPOX.js";
import "./session-file-rIikQBFx.js";
import { t as parseSessionThreadInfo } from "./thread-info-CfWpWYkw.js";
import "./transcript-BU7nLxj8.js";
import "./targets-CCzE5Q4H.js";
//#region src/config/sessions/main-session.runtime.ts
function resolveMainSessionKeyFromConfig() {
	return resolveMainSessionKey(loadConfig());
}
//#endregion
//#region src/config/sessions/delivery-info.ts
function extractDeliveryInfo(sessionKey) {
	const hasRoutableDeliveryContext = (context) => Boolean(context?.channel && context?.to);
	const { baseSessionKey, threadId } = parseSessionThreadInfo(sessionKey);
	if (!sessionKey || !baseSessionKey) return {
		deliveryContext: void 0,
		threadId
	};
	let deliveryContext;
	try {
		const store = loadSessionStore(resolveStorePath(loadConfig().session?.store));
		let entry = store[sessionKey];
		let storedDeliveryContext = deliveryContextFromSession(entry);
		if (!hasRoutableDeliveryContext(storedDeliveryContext) && baseSessionKey !== sessionKey) {
			entry = store[baseSessionKey];
			storedDeliveryContext = deliveryContextFromSession(entry);
		}
		if (hasRoutableDeliveryContext(storedDeliveryContext)) deliveryContext = {
			channel: storedDeliveryContext.channel,
			to: storedDeliveryContext.to,
			accountId: storedDeliveryContext.accountId,
			threadId: storedDeliveryContext.threadId != null ? String(storedDeliveryContext.threadId) : void 0
		};
	} catch {}
	return {
		deliveryContext,
		threadId
	};
}
//#endregion
export { resolveMainSessionKeyFromConfig as n, extractDeliveryInfo as t };
