// openclaw/dist/slash-dispatch.runtime-ZXEtCuuN.js
import { t as resolveConversationLabel$1 } from "./conversation-label-DGJN4j8D.js";
import { s as resolveChunkMode$1 } from "./chunk-BRF5Bnyz.js";
import { i as resolveAgentRoute$1 } from "./resolve-route-BWZgM2t0.js";
import "./routing-DSl48EkU.js";
import { t as finalizeInboundContext$1 } from "./inbound-context-xUb6mT6i.js";
import { n as dispatchReplyWithDispatcher$1 } from "./provider-dispatcher-RhkgL42J.js";
import { t as resolveMarkdownTableMode$1 } from "./markdown-tables-DgLnIvPZ.js";
import "./config-runtime-MrY1WAZr.js";
import "./reply-runtime-CHSEdiMI.js";
import { t as recordInboundSessionMetaSafe$1 } from "./conversation-runtime-DyA8YCJp.js";
import { r as deliverSlackSlashReplies$1 } from "./replies-DDk-0hqw.js";
//#region extensions/slack/src/monitor/slash-dispatch.runtime.ts
function resolveChunkMode(...args) {
	return resolveChunkMode$1(...args);
}
function finalizeInboundContext(...args) {
	return finalizeInboundContext$1(...args);
}
function dispatchReplyWithDispatcher(...args) {
	return dispatchReplyWithDispatcher$1(...args);
}
function resolveConversationLabel(...args) {
	return resolveConversationLabel$1(...args);
}
function recordInboundSessionMetaSafe(...args) {
	return recordInboundSessionMetaSafe$1(...args);
}
function resolveMarkdownTableMode(...args) {
	return resolveMarkdownTableMode$1(...args);
}
function resolveAgentRoute(...args) {
	return resolveAgentRoute$1(...args);
}
function deliverSlackSlashReplies(...args) {
	return deliverSlackSlashReplies$1(...args);
}
//#endregion
export { deliverSlackSlashReplies, dispatchReplyWithDispatcher, finalizeInboundContext, recordInboundSessionMetaSafe, resolveAgentRoute, resolveChunkMode, resolveConversationLabel, resolveMarkdownTableMode };
