// openclaw/dist/extensions/github-copilot/register.runtime.js
import { n as ensureAuthProfileStore } from "../../store-C6K7theB.js";
import { i as coerceSecretRef } from "../../types.secrets-D-R9X_jm.js";
import { n as listProfilesForProvider } from "../../profiles-BpuHvuWc.js";
import { r as resolveCopilotApiToken, t as DEFAULT_COPILOT_API_BASE_URL } from "../../github-copilot-token-eHXaFu56.js";
import "../../provider-auth-CUgkNKj0.js";
import { t as githubCopilotLoginCommand } from "../../provider-auth-login-DeJ8eryS.js";
import { n as resolveCopilotForwardCompatModel, t as PROVIDER_ID } from "../../models-DMMLDBPz.js";
import { n as wrapCopilotProviderStream, t as wrapCopilotAnthropicStream } from "../../stream-D3V_QLN-.js";
import "../../token-1uE3rue6.js";
import { t as fetchCopilotUsage } from "../../usage-Dix0Sj7m.js";
export { DEFAULT_COPILOT_API_BASE_URL, PROVIDER_ID, coerceSecretRef, ensureAuthProfileStore, fetchCopilotUsage, githubCopilotLoginCommand, listProfilesForProvider, resolveCopilotApiToken, resolveCopilotForwardCompatModel, wrapCopilotAnthropicStream, wrapCopilotProviderStream };
